matlab/src/matlab/+arrow/+tabular/Tabular.m

```diff
@@ -0,0 +1,125 @@
+% permissions and limitations under the License.
+
+classdef Tabular < matlab.mixin.CustomDisplay & matlab.mixin.Scalar
+
+    properties (Dependent, SetAccess=private, GetAccess=public)
+        NumRows
+        NumColumns
+        ColumnNames
+        Schema
+    end
+
+    properties (Hidden, SetAccess=private, GetAccess=public)
+        Proxy
+    end
+    
+    methods(Access=protected, Abstract)
+        % constructColumnFromProxy must construct an instance of the
+        % appropriate MATLAB class from the proxyInfo argument. The
+        % template method arrow.tabular.Tabular/column() invokes this
+        % method.
+        column = constructColumnFromProxy(obj, proxyInfo)
+    end
+
+    methods
+        function obj = Tabular(proxy)
+            arguments
+                proxy(1, 1) libmexclass.proxy.Proxy
+            end
+            import arrow.internal.proxy.validate
+            obj.Proxy = proxy;
+        end
+
+        function numColumns = get.NumColumns(obj)
+            numColumns = obj.Proxy.getNumColumns();
+        end
+
+        function numRows = get.NumRows(obj)
+            numRows = obj.Proxy.getNumRows();
+        end
+
+        function columnNames = get.ColumnNames(obj)
+            columnNames = obj.Proxy.getColumnNames();
+        end
+
+        function schema = get.Schema(obj)
+            proxyID = obj.Proxy.getSchema();
+            proxy = libmexclass.proxy.Proxy(Name="arrow.tabular.proxy.Schema", ID=proxyID);
+            schema = arrow.tabular.Schema(proxy);
+        end
+
+        function array = column(obj, idx)
+            import arrow.internal.validate.*
+
+            idx = index.numericOrString(idx, "int32", AllowNonScalar=false);
+
+            if isnumeric(idx)
+                args = struct(Index=idx);
+                proxyInfo = obj.Proxy.getColumnByIndex(args);
+            else
+                args = struct(Name=idx);
+                proxyInfo = obj.Proxy.getColumnByName(args);
+            end
+
+            array = obj.constructColumnFromProxy(proxyInfo);
+        end
+
+        function T = table(obj)
+            import arrow.tabular.internal.*
+
+            numColumns = obj.NumColumns;
+            matlabArrays = cell(1, numColumns);
+
+            for ii = 1:numColumns
+                matlabArrays{ii} = toMATLAB(obj.column(ii));
+            end
+
+            validVariableNames = makeValidVariableNames(obj.ColumnNames);
+            validDimensionNames = makeValidDimensionNames(validVariableNames);
+
+            T = table(matlabArrays{:}, ...
+                VariableNames=validVariableNames, ...
+                DimensionNames=validDimensionNames);
+        end
+
+        function T = toMATLAB(obj)
+            T = obj.table();
+        end
+
+        function tf = isequal(obj, varargin)
+            tf = arrow.tabular.internal.isequal(obj, varargin{:});
+        end
+
+    end
+
+    methods (Access = private)
+        function str = toString(obj)
```

Comment:
Maybe we should add a newline here so that the spacing is "balanced" inside of the `methods` block?

Reply:
done.